File: src/main/java/com/github/jactor/persistence/entity/GuestBookEntity.java
package com.github.jactor.persistence.entity;

import static java.util.Objects.hash;
import static java.util.stream.Collectors.toSet;

import com.github.jactor.persistence.dto.GuestBookDto;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import javax.persistence.AttributeOverride;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;

@Entity
@Table(name = "T_GUEST_BOOK")
public class GuestBookEntity implements PersistentEntity<GuestBookEntity> {

  @Id
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "guestBookSeq")
  @SequenceGenerator(name = "guestBookSeq", sequenceName = "T_GUEST_BOOK_SEQ", allocationSize = 1)
  private Long id;

  @Embedded
  @AttributeOverride(name = "createdBy", column = @Column(name = "CREATED_BY"))
  @AttributeOverride(name = "timeOfCreation", column = @Column(name = "CREATION_TIME"))
  @AttributeOverride(name = "modifiedBy", column = @Column(name = "UPDATED_BY"))
  @AttributeOverride(name = "timeOfModification", column = @Column(name = "UPDATED_TIME"))
  private PersistentDataEmbeddable persistentDataEmbeddable;

  @Column(name = "TITLE")
  private String title;
  @JoinColumn(name = "USER_ID")
  @OneToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE}, fetch = FetchType.LAZY)
  private UserEntity user;
  @OneToMany(mappedBy = "guestBook", cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
  private Set<GuestBookEntryEntity> entries = new HashSet<>();

  @SuppressWarnings("unused")
  GuestBookEntity() {
    // used by entity manager
  }

  /**
   * @param guestBook to copyWithoutId...
   */
  private GuestBookEntity(GuestBookEntity guestBook) {
    entries = guestBook.entries.stream().map(GuestBookEntryEntity::copyWithoutId).collect(toSet());
    id = guestBook.id;
    persistentDataEmbeddable = new PersistentDataEmbeddable();
    title = guestBook.title;
    user = guestBook.copyUserWithoutId();
  }

  public GuestBookEntity(@NotNull GuestBookDto guestBook) {
    entries = guestBook.getEntries().stream().map(GuestBookEntryEntity::new).collect(toSet());
    id = guestBook.getId();
    persistentDataEmbeddable = new PersistentDataEmbeddable(guestBook.fetchPersistentDto());
    title = guestBook.getTitle();
    user = Optional.ofNullable(guestBook.getUser()).map(UserEntity::new).orElse(null);
  }

  private UserEntity copyUserWithoutId() {
    return Optional.ofNullable(user).map(UserEntity::copyWithoutId).orElse(null);
  }

  public GuestBookDto asDto() {
    return new GuestBookDto(
        persistentDataEmbeddable.asPersistentDto(id),
        entries.stream().map(GuestBookEntryEntity::asDto).collect(toSet()),
        title,
        Optional.ofNullable(user).map(UserEntity::asDto).orElse(null)
    );
  }

  @Override
  public GuestBookEntity copyWithoutId() {
    GuestBookEntity guestBookEntity = new GuestBookEntity(this);
    guestBookEntity.setId(null);

    return guestBookEntity;
  }

  @Override
  public void modify() {
    persistentDataEmbeddable.modify();
  }

  public void add(GuestBookEntryEntity guestBookEntry) {
    entries.add(guestBookEntry);
    guestBookEntry.setGuestBook(this);
  }

  @Override
  public boolean equals(Object o) {
    return this == o || o != null && getClass() == o.getClass() &&
        Objects.equals(getTitle(), ((GuestBookEntity) o).getTitle()) &&
        Objects.equals(getUser(), ((GuestBookEntity) o).getUser());
  }

  @Override
  public int hashCode() {
    return hash(title, user);
  }

  @Override
  public String toString() {
    return new ToStringBuilder(this, ToStringStyle.SIMPLE_STYLE)
        .appendSuper(super.toString())
        .append(getTitle())
        .append(getUser())
        .toString();
  }

  @Override
  public String getCreatedBy() {
    return persistentDataEmbeddable.getCreatedBy();
  }

  @Override
  public LocalDateTime getTimeOfCreation() {
    return persistentDataEmbeddable.getTimeOfCreation();
  }

  @Override
  public String getModifiedBy() {
    return persistentDataEmbeddable.getModifiedBy();
  }

  @Override
  public LocalDateTime getTimeOfModification() {
    return persistentDataEmbeddable.getTimeOfModification();
  }

  @Override
  public Long getId() {
    return id;
  }

  @Override
  public void setId(Long id) {
    this.id = id;
  }

  public Set<GuestBookEntryEntity> getEntries() {
    return entries;
  }

  public String getTitle() {
    return title;
  }

  public UserEntity getUser() {
    return user;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public void setUser(UserEntity user) {
    this.user = user;
  }

  public static GuestBookEntity aGuestBook(GuestBookDto guestBookDto) {
    return new GuestBookEntity(guestBookDto);
  }
}

File: persistence/src/main/kotlin/com/github/jactor/persistence/entity/GuestBookEntity.kt
package com.github.jactor.persistence.entity

import java.time.LocalDateTime
import java.util.Objects
import org.apache.commons.lang3.builder.ToStringBuilder
import org.apache.commons.lang3.builder.ToStringStyle
import com.github.jactor.persistence.dto.GuestBookDto
import jakarta.persistence.AttributeOverride
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToMany
import jakarta.persistence.OneToOne
import jakarta.persistence.SequenceGenerator
import jakarta.persistence.Table

@Entity
@Table(name = "T_GUEST_BOOK")
class GuestBookEntity : PersistentEntity<GuestBookEntity?> {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "guestBookSeq")
    @SequenceGenerator(name = "guestBookSeq", sequenceName = "T_GUEST_BOOK_SEQ", allocationSize = 1)
    override var id: Long? = null

    @Embedded
    @AttributeOverride(name = "createdBy", column = Column(name = "CREATED_BY"))
    @AttributeOverride(name = "timeOfCreation", column = Column(name = "CREATION_TIME"))
    @AttributeOverride(name = "modifiedBy", column = Column(name = "UPDATED_BY"))
    @AttributeOverride(name = "timeOfModification", column = Column(name = "UPDATED_TIME"))
    private lateinit var persistentDataEmbeddable: PersistentDataEmbeddable

    @Column(name = "TITLE")
    var title: String? = null

    @JoinColumn(name = "USER_ID")
    @OneToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE], fetch = FetchType.LAZY)
    var user: UserEntity? = null

    @OneToMany(mappedBy = "guestBook", cascade = [CascadeType.MERGE], fetch = FetchType.EAGER)
    private var entries: MutableSet<GuestBookEntryEntity> = HashSet()

    constructor() {
        // used by entity manager
    }

    /**
     * @param guestBook to copyWithoutId...
     */
    private constructor(guestBook: GuestBookEntity) {
        entries = guestBook.entries.map { it.copyWithoutId() }.toMutableSet()
        id = guestBook.id
        persistentDataEmbeddable = PersistentDataEmbeddable()
        title = guestBook.title
        user = guestBook.copyUserWithoutId()
    }

    constructor(guestBook: GuestBookDto) {
        entries = guestBook.entries.map { GuestBookEntryEntity(it) }.toMutableSet()
        id = guestBook.id
        persistentDataEmbeddable = PersistentDataEmbeddable(guestBook.persistentDto)
        title = guestBook.title
        user = guestBook.userInternal?.let { UserEntity(it) }
    }

    private fun copyUserWithoutId(): UserEntity? {
        return user?.copyWithoutId()
    }

    fun asDto(): GuestBookDto {
        return GuestBookDto(
            persistentDto = persistentDataEmbeddable.asPersistentDto(id),
            entries = entries.map { it.asDto() }.toMutableSet(),
            title = title,
            userInternal = user?.asDto()
        )
    }

    override fun copyWithoutId(): GuestBookEntity {
        val guestBookEntity = GuestBookEntity(this)
        guestBookEntity.id = null
        return guestBookEntity
    }

    override fun modifiedBy(modifier: String): GuestBookEntity {
        persistentDataEmbeddable.modifiedBy(modifier)
        return this
    }

    fun add(guestBookEntry: GuestBookEntryEntity) {
        entries.add(guestBookEntry)
        guestBookEntry.guestBook = this
    }

    override fun equals(other: Any?): Boolean {
        return this === other || other != null && javaClass == other.javaClass &&
            title == (other as GuestBookEntity).title &&
            user == other.user
    }

    override fun hashCode(): Int {
        return Objects.hash(title, user)
    }

    override fun toString(): String {
        return ToStringBuilder(this, ToStringStyle.SIMPLE_STYLE)
            .appendSuper(super.toString())
            .append(title)
            .append(user)
            .toString()
    }

    override val createdBy: String
        get() = persistentDataEmbeddable.createdBy
    override val timeOfCreation: LocalDateTime
        get() = persistentDataEmbeddable.timeOfCreation
    override val modifiedBy: String
        get() = persistentDataEmbeddable.modifiedBy
    override val timeOfModification: LocalDateTime
        get() = persistentDataEmbeddable.timeOfModification

    fun getEntries(): Set<GuestBookEntryEntity> {
        return entries
    }

    companion object {
        @JvmStatic
        fun aGuestBook(guestBookDto: GuestBookDto): GuestBookEntity {
            return GuestBookEntity(guestBookDto)
        }
    }
}
